engine/access/rest/routes/websocket_handler.go

```diff
@@ -0,0 +1,279 @@
+	if err != nil {
+		return models.NewRestError(http.StatusInternalServerError, "Set the initial write deadline error: ", err)
+	}
+	err = wsController.conn.SetReadDeadline(time.Now().Add(pongWait)) // Set the initial read deadline for the first pong message
+	if err != nil {
+		return models.NewRestError(http.StatusInternalServerError, "Set the initial read deadline error: ", err)
+	}
+	// Establish a Pong handler
+	wsController.conn.SetPongHandler(func(string) error {
+		err := wsController.conn.SetReadDeadline(time.Now().Add(pongWait))
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	return nil
+}
+
+// wsErrorHandler handles WebSocket errors by sending an appropriate close message
+// to the client WebSocket connection.
+//
+// If the error is an instance of models.StatusError, the function extracts the
+// relevant information like status code and user message to construct the WebSocket
+// close code and message. If the error is not a models.StatusError, a default
+// internal server error close code and the error's message are used.
+// The connection is then closed using WriteControl to send a CloseMessage with the
+// constructed close code and message. Any errors that occur during the closing
+// process are logged using the provided logger.
+func (wsController *WebsocketController) wsErrorHandler(err error) {
+	// rest status type error should be returned with status and user message provided
+	var statusErr models.StatusError
+	var wsCode int
+	var wsMsg string
+
+	if errors.As(err, &statusErr) {
+		if statusErr.Status() == http.StatusBadRequest {
+			wsCode = websocket.CloseUnsupportedData
+		}
+		if statusErr.Status() == http.StatusServiceUnavailable {
+			wsCode = websocket.CloseTryAgainLater
+		}
+		if statusErr.Status() == http.StatusRequestTimeout {
+			wsCode = websocket.CloseGoingAway
+		}
+		wsMsg = statusErr.UserMessage()
+
+	} else {
+		wsCode = websocket.CloseInternalServerErr
+		wsMsg = err.Error()
+	}
+
+	// Close the connection with the CloseError message
+	err = wsController.conn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(wsCode, wsMsg), time.Now().Add(time.Second))
+	if err != nil {
+		wsController.logger.Error().Err(err).Msg(fmt.Sprintf("error sending WebSocket error: %v", err))
+	}
+}
+
+// writeEvents use for writes events and pings to the WebSocket connection for a given subscription.
+// It listens to the subscription's channel for events and writes them to the WebSocket connection.
+// If an error occurs or the subscription channel is closed, it handles the error or termination accordingly.
+// The function uses a ticker to periodically send ping messages to the client to maintain the connection.
+func (wsController *WebsocketController) writeEvents(sub state_stream.Subscription) {
+	ticker := time.NewTicker(pingPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case _, ok := <-wsController.readChannel:
+			if !ok {
```

Comment:
```suggestion
                        // we use `readChannel` as indicator of client's status, when `readChannel` closes it means that client connection has been terminated and we need to stop this goroutine to avoid memory leak.
			if !ok {
```